File: Exercism Codes/Java/BaseConverter.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;

class BaseConverter {

    int base;
    int[] numbers;

    BaseConverter(int base, int[] numbers) {
        if(base < 2) throw new IllegalArgumentException("Bases must be at least 2.");
        if(Arrays.stream(numbers).anyMatch(n -> n >= base))
            throw new IllegalArgumentException("All digits must be strictly less than the base.");
        if(Arrays.stream(numbers).anyMatch(n -> n < 0))
            throw new IllegalArgumentException("Digits may not be negative.");
        this.base = base;
        this.numbers = numbers;
    }

    List<Integer> convertToBase(int base) {
        if(base < 2) throw new IllegalArgumentException("Bases must be at least 2.");
        int dec = IntStream.range(0, numbers.length)
                .reduce(0, (acc, next) -> acc + (numbers[numbers.length - 1 - next] * (int) Math.pow(this.base, next)));
        List<Integer> nums = new ArrayList<>();
        do {
            nums.add(0, dec % base);
            dec /= base;
        }
        while(dec > 0);
        return nums;
    }
}

File: Exercism Codes/Kotlin/BaseConverter.kt
import kotlin.math.pow

class BaseConverter(private val base: Int, private val numbers: IntArray) {

    init {
        require(base >= 2) { "Bases must be at least 2." }
        require(numbers.isNotEmpty()) { "You must supply at least one digit." }
        require(!(numbers.size > 1 && numbers[0] == 0)) { "Digits may not contain leading zeros." }
        require(numbers.all { it >= 0 }) { "Digits may not be negative." }
        require(numbers.all { it < base }) { "All digits must be strictly less than the base." }
    }

    fun convertToBase(newBase: Int): IntArray {
        require(newBase >= 2) { "Bases must be at least 2." }
        var base10 = numbers
            .mapIndexed { idx, num -> num * base.toDouble().pow(numbers.size - 1.0 - idx).toInt() }
            .sum()
        val result = mutableListOf<Int>()
        do {
            result.add(0, base10 % newBase)
            base10 /= newBase
        } while (base10 > 0)
        return result.toIntArray()
    }
}
